pertpy/tools/_distances.py

```diff
@@ -0,0 +1,230 @@
+            Y (np.ndarray): Second vector of shape (n_samples, n_features).
+
+        Returns:
+            float: Distance between X and Y.
+        """
+        raise NotImplementedError("Metric class is abstract.")
+
+    @abstractmethod
+    def from_precomputed(self, P: np.ndarray, idx: np.ndarray, **kwargs) -> float:
+        """Compute a distance between vectors X and Y with precomputed distances.
+
+        Args:
+            P (np.ndarray): Pairwise distance matrix of shape (n_samples, n_samples).
+            idx (np.ndarray): Boolean array of shape (n_samples,) indicating which
+            samples belong to X (or Y, since each metric is symmetric).
+
+        Returns:
+            float: Distance between X and Y.
+        """
+        raise NotImplementedError("Metric class is abstract.")
+
+
+# Specific distance metrics
+class Edistance(AbstractDistance):
+    """Edistance metric."""
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.accepts_precomputed = True
+
+    def __call__(self, X: np.ndarray, Y: np.ndarray, **kwargs) -> float:
+        # TODO: inherit docstring from parent class
+        sigma_X = pairwise_distances(X, X, metric="euclidean").mean()
+        sigma_Y = pairwise_distances(Y, Y, metric="euclidean").mean()
+        delta = pairwise_distances(X, Y, metric="euclidean").mean()
+        return 2 * delta - sigma_X - sigma_Y
+
+    def from_precomputed(self, P: np.ndarray, idx: np.ndarray, **kwargs) -> float:
+        # TODO: inherit docstring from parent class
+        sigma_X = P[idx, :][:, idx].mean()
+        sigma_Y = P[~idx, :][:, ~idx].mean()
+        delta = P[idx, :][:, ~idx].mean()
+        return 2 * delta - sigma_X - sigma_Y
+
+
+# class MMD(AbstractDistance):
+#     # TODO: implement MMD metric
+#     # e.g.
+#     # https://github.com/jindongwang/transferlearning/blob/master/code/distance/mmd_numpy_sklearn.py
+#     # https://github.com/calico/scmmd
+#     pass
+
+# class WassersteinDistance(AbstractDistance):
+#     """Wasserstein distance metric.
+#     NOTE: I bet the Theislab OT people have a better solution for this.
+#     """
+#     # TODO: implement Wasserstein distance metric
+
+
+class PseudobulkDistance(AbstractDistance):
+    """Euclidean distance between pseudobulk vectors."""
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.accepts_precomputed = False
+
+    def __call__(self, X: np.ndarray, Y: np.ndarray, **kwargs) -> float:
+        return np.linalg.norm(X.mean(axis=0) - Y.mean(axis=0), ord=2, **kwargs)
+
+    def from_precomputed(self, P: np.ndarray, idx: np.ndarray, **kwargs) -> float:
+        raise NotImplementedError("PseudobulkDistance cannot be called on a pairwise distance matrix.")
+
+
+class MeanPairwiseDistance(AbstractDistance):
+    """Mean of the pairwise euclidean distance between two groups of cells."""
+
+    # NOTE: I think this might be basically the same as PseudobulkDistance
```

Comment:
So what's the difference? If there's any? If there's not we can remove this.

Reply:
I was not entirely sure how the mean of pairwise distances relates the pairwise distance between means. Empirically, they are not the same. So let's keep and I'll move the comment from code back to my head ^^